src/lib.rs

```diff
@@ -647,6 +694,285 @@ impl Bump {
+    ///
+    /// Even when optimizations are on, this function does not **guarantee**
+    /// that the value is constructed on the heap. To the best of our
+    /// knowledge no such guarantee can be made in stable Rust as of 1.44.
+    ///
+    /// Rewinding in the `Err` case will fail if `f` makes any additional
+    /// allocations in `self`.
+    ///
+    /// ## Errors
+    ///
+    /// Iff the allocation succeeds but `f` fails, that error is forwarded by value.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if reserving space for `Result<T, E>` would cause an overflow.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let bump = bumpalo::Bump::new();
+    /// let x = bump.alloc_try_with(|| Ok("hello"))?;
+    /// assert_eq!(*x, "hello");
+    /// # Result::<_, ()>::Ok(())
+    /// ```
+    #[inline(always)]
+    #[allow(clippy::mut_from_ref)]
+    pub fn alloc_try_with<F, T, E>(&self, f: F) -> Result<&mut T, E>
+    where
+        F: FnOnce() -> Result<T, E>,
+        E: Unpin,
+    {
+        let rewind_footer = self.current_chunk_footer.get();
+        let rewind_ptr = unsafe { rewind_footer.as_ref() }.ptr.get();
+        let ptr = NonNull::from(self.alloc_with(f));
+        match unsafe { ptr.clone().as_mut() } {
+            Ok(t) => Ok(unsafe {
+                //SAFETY:
+                // The `&mut Result<T, E>` returned by `alloc_with` may be
+                // lifetime-limited by `E`, but the derived `&mut T` still has
+                // the same validity as in `alloc_with` since the error variant
+                // is already ruled out here.
+
+                // We could conditionally truncate the allocation here, but since
+                // it grows backwards, it seems unlikely that we'd get any more
+                // than the `Result`'s discriminant this way, if anything at all.
+                &mut *(t as *mut _)
+            }),
+            Err(e) => unsafe {
+                //SAFETY:
+                // As `E: Unpin`, we can just copy the value here as long as we
+                // avoid a double-drop (which can't happen as any specific
+                // references to the `E`'s data in `self` are destroyed when
+                // this function returns).
+                let current_footer_p = self.current_chunk_footer.get();
+                let current_ptr = &current_footer_p.as_ref().ptr;
+                if current_ptr.get() == ptr.cast() {
+                    // We can also reuse the memory, unless `f` made any
+                    // further allocations in `self`.
+                    if current_footer_p == rewind_footer {
+                        current_ptr.set(rewind_ptr)
+                    } else {
+                        // If the current chunk changed, we can at least reset
+                        // to its start, since we know no other allocations
+                        // happened.
+                        current_ptr.set(current_footer_p.as_ref().data)
```

Comment:
> For example, it could have done something like
> 
> ```rust
> let x: Option<&mut u32> = None;
> let _ = b.alloc_try_with(|| {
>     x = Some(b.alloc(42));
>     Err(())
> });
> x.unwrap()
> ```

This will make the `current_ptr.get() == ptr.cast()` check fail, regardless of the chunk change. I've clarified the comments regarding that.

(Please bear with me while I go through these one by one. I'll squash and rebase everything afterwards.)

Reply:
> It seems like all this hairy unsafe pointer code could be removed by reusing the `<Bump as Alloc>::dealloc` method here, which already makes sure it only actually adjusts the bump pointer if the given pointer points to the last allocation made.

The reason I didn't use `dealloc` is that that method won't free alignment bytes, even though we have the necessary `rewind_ptr` here to do this.

If leaking them is okay, I'll change it, but I personally would prefer to do this as cleanly (in that regard) as possible.